File: javascript/24.两两交换链表中的节点.js
/*
 * @lc app=leetcode.cn id=24 lang=javascript
 *
 * [24] 两两交换链表中的节点
 * 
 * 交换链表中节点的题，只要理清楚和用变量保存清楚 pre -> a -> b -> tail的关系即可
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    if (!head || !head.next) { return head; }

    var pre = new ListNode();
    pre.next = head;
    var move = head;
    var res = head.next;
    while (move && move.next) {
        var a = move;
        var b = move.next;
        var tail = move.next.next;

        pre.next = b;
        b.next = a;
        a.next = tail;

        pre = move;
        move = move.next;
    }
    return res;
};
// @lc code=end

File: java/24.两两交换链表中的节点.java
/*
 * @lc app=leetcode.cn id=24 lang=java
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head==null || head.next==null) { return head; }

        ListNode newHead = new ListNode(0, head);
        ListNode pre=newHead;
        ListNode move = head;
        while (move!=null && move.next!=null) {
            ListNode tail = move.next.next;
            ListNode a = move;
            pre.next = a.next;
            a.next.next = a;
            a.next = tail;
            
            move = tail;
            pre = a;
        }

        return newHead.next;
    }
}
// @lc code=end

